File: utils/phone.py
"""
BharatConnect Canonical Phone Normalization Utility (utils/phone.py)

Provides consistent, robust E.164 and national phone number parsing, validation,
and normalization across all backend services, authentication flows, and contact matching.
"""

import re
from typing import Optional, Dict

def normalize_phone(phone: Optional[str], default_country_code: str = "91") -> str:
    """
    Normalizes input phone string into a clean 10-digit national canonical string.
    Strips non-digits, country code prefixes (e.g. +91, 91), and leading zeros.
    
    Examples:
      '+91 9876543210' -> '9876543210'
      '09876543210'    -> '9876543210'
      '919876543210'   -> '9876543210'
      '98765 43210'    -> '9876543210'
    """
    if not phone:
        return ""
    
    raw = str(phone).strip()
    digits = re.sub(r"\D", "", raw)
    
    if not digits:
        return ""
    
    # Remove leading zeros
    digits = digits.lstrip("0")
    
    # If 12 digits starting with country code '91', strip country code
    if len(digits) == 12 and digits.startswith("91"):
        digits = digits[2:]
    elif len(digits) > 10 and digits.startswith(default_country_code):
        digits = digits[len(default_country_code):]
    
    # If digits exceed 10, extract trailing 10 digits for Indian standard phone numbers
    if len(digits) > 10:
        digits = digits[-10:]
        
    return digits


def to_e164(phone: Optional[str], country_code: str = "+91") -> str:
    """
    Formats a phone number to official E.164 format (e.g., '+919876543210').
    """
    normalized = normalize_phone(phone)
    if not normalized:
        return ""
    prefix = country_code if country_code.startswith("+") else f"+{country_code}"
    return f"{prefix}{normalized}"


def validate_phone(phone: Optional[str]) -> bool:
    """
    Validates whether the input string represents a valid 10-digit mobile number.
    """
    normalized = normalize_phone(phone)
    return len(normalized) == 10 and normalized[0] in "6789"


def parse_phone(phone: Optional[str]) -> Dict[str, str]:
    """
    Returns full metadata dictionary for a given phone number.
    """
    norm = normalize_phone(phone)
    is_valid = validate_phone(phone)
    e164 = to_e164(phone) if is_valid else ""
    
    return {
        "raw": str(phone or ""),
        "normalized": norm,
        "e164": e164,
        "is_valid": is_valid
    }

File: android_app/app/src/main/assets/www/utils/phone.js
/**
 * BharatConnect Client Canonical Phone Normalization Utility (www/utils/phone.js)
 */

window.BharatConnectPhone = {
    /**
     * Normalizes input phone string into a clean 10-digit national canonical string.
     * Strips non-digits, country code prefixes (+91, 91), and leading zeros.
     * 
     * @param {string} phone
     * @param {string} defaultCountryCode
     * @returns {string} 10-digit canonical phone string
     */
    normalizePhone: function(phone, defaultCountryCode = '91') {
        if (!phone) return '';
        let digits = String(phone).replace(/\D/g, '');
        if (!digits) return '';

        digits = digits.replace(/^0+/, '');

        if (digits.length === 12 && digits.startsWith('91')) {
            digits = digits.slice(2);
        } else if (digits.length > 10 && digits.startsWith(defaultCountryCode)) {
            digits = digits.slice(defaultCountryCode.length);
        }

        if (digits.length > 10) {
            digits = digits.slice(-10);
        }

        return digits;
    },

    /**
     * Formats phone number to E.164 string format (e.g., '+919876543210')
     * @param {string} phone 
     * @param {string} countryCode 
     * @returns {string}
     */
    toE164: function(phone, countryCode = '+91') {
        const norm = this.normalizePhone(phone);
        if (!norm) return '';
        const prefix = countryCode.startsWith('+') ? countryCode : '+' + countryCode;
        return prefix + norm;
    },

    /**
     * Validates whether string is a valid 10-digit Indian mobile number
     * @param {string} phone 
     * @returns {boolean}
     */
    validatePhone: function(phone) {
        const norm = this.normalizePhone(phone);
        return norm.length === 10 && ['6', '7', '8', '9'].includes(norm[0]);
    }
};
